lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,225 @@
+      throw new ERR_INVALID_ARG_VALUE(
+        'init',
+        object,
+        'is not a valid array entry'
+      );
+    }
+  } else if (!isBoxedPrimitive(object)) {
+    // Object of key/value entries
+    const entries = ObjectEntries(object);
+    for (let i = 0; i < entries.length; i++) {
+      headers.append(entries[i][0], entries[i][1]);
+    }
+  }
+}
+
+class Headers {
+  constructor(init) {
+    this[kHeadersList] = [];
+
+    if (init &&
+      validateObject(init, 'init', { allowArray: true }) === undefined) {
+      fill(this, init);
+    }
+  }
+
+  append(name, value) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] += `, ${normalizedValue}`;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  delete(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList].splice(index, 2);
+    }
+  }
+
+  get(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      return this[kHeadersList][index + 1];
+    }
+
+    return null;
+  }
+
+  has(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    return this[kHeadersList][index] === normalizedName;
+  }
+
+  set(name, value) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] = normalizedValue;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  * keys() {
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      yield this[kHeadersList][index];
+    }
+  }
+
+  * values() {
+    for (let index = 1; index < this[kHeadersList].length; index += 2) {
+      yield this[kHeadersList][index];
+    }
+  }
+
+  * entries() {
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      yield [this[kHeadersList][index], this[kHeadersList][index + 1]];
+    }
+  }
+
+  forEach(callback, thisArg) {
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      callback.call(
+        thisArg,
+        this[kHeadersList][index + 1],
+        this[kHeadersList][index],
+        this
+      );
+    }
+  }
```

Comment:
Should implement a custom inspect function in here so that the kHeadersList internal property is not directly exposed.

Reply:
Not sure what you mean here. Do you have an example of "custom inspect function" I can read through?

Reply:
Do you mean something like:

```js
const { inspect } = require('internal/util/inspect');

class Headers {
  [inspect.custom] () {
    return this[kHeadersList];
  }
}
```

Reply:
Take a look in `lib/internal/webstreams/readablestreams` for an example. Look for the `[kInspect]` function.